src/index.js

```diff
@@ -0,0 +1,185 @@
+import Bowser from 'bowser/src/bowser'
+
+/** @enum {string} */
+const ONBOARDING_STATE = {
+  INSTALLED: 'INSTALLED',
+  NOT_INSTALLED: 'NOT_INSTALLED',
+  REGISTERED: 'REGISTERED',
+  REGISTERING: 'REGISTERING',
+  RELOADING: 'RELOADING',
+}
+
+/** @enum {string} */
+const EXTENSION_DOWNLOAD_URL = {
+  CHROME: 'https://chrome.google.com/webstore/detail/metamask/nkbihfbeogaeaoehlefnkodbefgpgknn',
+  FIREFOX: 'https://addons.mozilla.org/en-US/firefox/addon/ether-metamask/',
+  DEFAULT: 'https://metamask.io',
+}
+
+/** @enum {string} */
+const FORWARDER_MODE = {
+  INJECT: 'INJECT',
+  OPEN_TAB: 'OPEN_TAB',
+}
+
+// sessionStorage key
+const REGISTRATION_IN_PROGRESS = 'REGISTRATION_IN_PROGRESS'
+
+// forwarder iframe id
+const FORWARDER_ID = 'FORWARDER_ID'
+
+
+/**
+ * @typedef {Object} OnboardingOptions - Options for configuring onboarding
+ * @property {string} [forwarderOrigin] - The origin of the forwarder page
+ * @property {string} [forwarderMode] - The method used for opening the forwarder ('TAB' or 'INJECT')
+ */
+
+class Onboarding {
+  /**
+   *
+   * @param {OnboardingOptions} [options] - Options for configuring onboarding
+   */
+  constructor ({ forwarderOrigin = 'https://fwd.metamask.io', forwarderMode = FORWARDER_MODE.INJECT } = {}) {
+    this.forwarderOrigin = forwarderOrigin
+    this.forwarderMode = forwarderMode
+    this.state = Onboarding.isMetaMaskInstalled() ?
+      ONBOARDING_STATE.INSTALLED :
+      ONBOARDING_STATE.NOT_INSTALLED
+
+    const browser = Onboarding._detectBrowser()
+    if (browser) {
+      this.downloadUrl = EXTENSION_DOWNLOAD_URL[browser]
+    } else {
+      this.downloadUrl = EXTENSION_DOWNLOAD_URL['DEFAULT']
+    }
+
+    this._onMessage = this._onMessage.bind(this)
+    this._onMessageFromForwarder = this._onMessageFromForwarder.bind(this)
+    this._openForwarder = this._openForwarder.bind(this)
+    this._openDownloadPage = this._openDownloadPage.bind(this)
+    this.startOnboarding = this.startOnboarding.bind(this)
+    this.stopOnboarding = this.stopOnboarding.bind(this)
+
+    window.addEventListener('message', this._onMessage.bind(this), false)
```

Comment:
Nit: `useCapture` defaults to `false` and we could use the reference to `this._onMessage` that's bound above

```suggestion
    window.addEventListener('message', this._onMessage)
```